neural_lam/datastore/mllam.py

```diff
@@ -0,0 +1,377 @@
+        `{category}_std`
+        variable for each
+        variable in the
+        category. For
+        `category=="state"`,
+        the dataarray should
+        also contain a
+        `state_diff_mean` and
+        `state_diff_std`
+        variable for the one-
+        step differences of
+        the state variables.
+
+        Parameters
+        ----------
+        category : str
+            The category of the dataset (state/forcing/static).
+
+        Returns
+        -------
+        xr.Dataset
+            The normalization dataarray for the given category, with variables
+            for the mean and standard deviation of the variables (and
+            differences for state variables).
+
+        """
+        ops = ["mean", "std"]
+        split = "train"
+        stats_variables = {
+            f"{category}__{split}__{op}": f"{category}_{op}" for op in ops
+        }
+        if category == "state":
+            stats_variables.update(
+                {f"state__{split}__diff_{op}": f"state_diff_{op}" for op in ops}
+            )
+
+        ds_stats = self._ds[stats_variables.keys()].rename(stats_variables)
+        return ds_stats
+
+    @property
+    def boundary_mask(self) -> xr.DataArray:
+        """Produce a 0/1 mask
+        for the boundary
+        points of the dataset,
+        these will sit at the
+        edges of the domain
+        (in x/y extent) and
+        will be used to mask
+        out the boundary
+        points from the loss
+        function and to
+        overwrite the boundary
+        points from the
+        prediction. For now
+        this is created when
+        the mask is requested,
+        but in the future this
+        could be saved to the
+        zarr file.
+
+        Returns
+        -------
+        xr.DataArray
+            A 0/1 mask for the boundary points of the dataset, where 1 is a
+            boundary point and 0 is not.
+
+        """
+        ds_unstacked = self.unstack_grid_coords(da_or_ds=self._ds)
+        da_state_variable = ds_unstacked["state"].isel(time=0).isel(state_feature=0)
+        da_domain_allzero = xr.zeros_like(da_state_variable)
+        ds_unstacked["boundary_mask"] = da_domain_allzero.isel(
+            x=slice(self._n_boundary_points, -self._n_boundary_points),
+            y=slice(self._n_boundary_points, -self._n_boundary_points),
+        )
+        ds_unstacked["boundary_mask"] = ds_unstacked.boundary_mask.fillna(1).astype(int)
+        return self.stack_grid_coords(da_or_ds=ds_unstacked.boundary_mask)
+
+    @property
+    def coords_projection(self) -> ccrs.Projection:
+        """Return the projection of the coordinates.
+
+        Returns
+        -------
+        ccrs.Projection
+            The projection of the coordinates.
+
+        """
+        # TODO: danra doesn't contain projection information yet, but the next
+        # version will for now we hardcode the projection
+        # XXX: this is wrong
```

Comment:
Having projection information with the data is nice. However, I would probably flip the priority order: have the user-specified take priority over what is in dataset. The use cases for user-specified are:

1. There is no projection specified in the dataset (no priority order needed)
2. There is something wrong with the projection in the dataset, fixes include:
  a. Fix projection in dataset (no priority order needed)
  b. Specify projection manually (priority must be on user-specified over dataset)

Reply:
I wasn't sure where to add this information. On the one hand I would like `mllam-data-prep` to actually understand and use the projection information (for example be able to compare datasets with different projections, parse and write projection information) but for now that seems like it needs some thought and we need to keep moving.

So what I have done is add a new section to the `mllam-data-prep` config that I have called the `extra` section (https://github.com/mllam/mllam-data-prep/pull/18). In here the user can add anything they want, `mllam-data-prep` doesn't use and doesn't validate anything in that part of the config. We can then use that to include the projection info for now and rather than reading projection info from the zarr file it is now parsed from the config that `mllam-data-prep` creates the config from. https://github.com/leifdenby/neural-lam/commit/67998b836345336dae4514786c027d3ac1d10347#diff-2ad89866cd8911c976c4758935816347911e858a77742c3f4d39a9f30f95368aR90. This of course assumes that all the x,y values in a given zarr dataset created with mllam-data-prep are in the same projection. The good thing is that `mllam-data-prep` checks that the coordinate values are coincident when combining different datasets, so it would be a weird fluke if the coordinates didn't overlap. I think this is ok for now, hopefully you both think so too. 

The capability for allowing this new `extra` section in the `mllam-data-prep` config is in a PR at the moment. I will create a new release before merging this PR, but now my branch with this new feature is what is referred to in `pyproject.toml`